add elasticsearch hosting

Alirexaa opened the pull request:
- Add Elasticsearch hosting resource
close: #4418
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4430)


- [x] [Aspire.Hosting.Elasticsearch](https://github.com/Alirexaa/aspire/tree/elasticsearch-hosing/src/Aspire.Hosting.Elasticsearch)
- [x] Aspire.Elastic.Clients.Elasticsearch

Alirexaa committed:
add elasticsearch hosting

Alirexaa commented:
> @dotnet-policy-service agree

Alirexaa commented:
@dotnet-policy-service agree

davidfowl commented on playground/Elasticseach/Elasticseach.ApiService/Program.cs:
```diff
@@ -0,0 +1,39 @@
+
+builder.AddServiceDefaults();
+
+builder.Services.AddSingleton<ElasticLowLevelClient>((sp) =>
```
This should be encapsulated in a component right? See https://github.com/dotnet/aspire/tree/main/src/Components#readme

Alirexaa replied:
@davidfowl 
for encapsulation code in a component, there are multiple packages like Nest ,Elasticsearch.Net and Elastic.Clients.Elasticsearch. The two first packages are deprecated and only **Elastic.Clients.Elasticsearch** is currently active. 

Elastic.Clients.Elasticsearch package is not in the package sources. only Elasticsearch.Net is available but already deprecated.

So, Isn't that a bad approach to using a deprecated client?
I think it's better Elastic.Clients.Elasticsearch package added in package sources.

davidfowl replied:
Sure. While we're looking for the right client, you can change the nuget.config to add NuGet.org so you can pull it in. We'll mirror the package once we decide what to use.

 Elastic.Clients.Elasticsearch seems like the one!

https://www.elastic.co/guide/en/elasticsearch/client/net-api/current/installation.html

mitchdenny replied:
Running the package import for `8.14.0` of `Elastic.Clients.Elasticsearch` now.

mitchdenny replied:
Its imported.

davidfowl commented on src/Aspire.Hosting.Elasticsearch/ElasticsearchResource.cs:
```diff
@@ -0,0 +1,40 @@
+/// A resource that represents a Elasticsearch
+/// </summary>
+/// <param name="name">The name of the resource.</param>
+public class ElasticsearchResource(string name) : ContainerResource(name), IResourceWithConnectionString
```
Consider looking into password support by default. We should be password protecting all of our resources by default.

Alirexaa replied:
@davidfowl 
Add password support in the [commit](https://github.com/dotnet/aspire/pull/4430/commits/ca9709c1a15ec37768652c4cb35be64692bd26b8).

I only used the password as a resource parameter in the constructor because it is not possible to change the built-in elastic superuser name, but I added the username as a property in the resource to have access to the username property out of class (for different authentication way in clients)

Alirexaa committed:
Add password support to elasticsearch resource

mitchdenny commented on Aspire.sln:
```diff
@@ -1,4 +1,5 @@
-﻿Microsoft Visual Studio Solution File, Format Version 12.00
+﻿
```
Why the extra line above here?

mitchdenny commented on Aspire.sln:
```diff
@@ -487,18 +488,22 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Aspire.Milvus.Client.Tests"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Elasticsearch", "Elasticsearch", "{5C28905E-0EE2-41C2-A161-7400671E66F7}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Elasticseach.AppHost", "playground\Elasticseach\Elasticseach.AppHost\Elasticseach.AppHost.csproj", "{4926BEFF-6A63-422F-8758-BB8EBEDD6BE8}"
```
`Elasticseach` -> `Elasticsearch`

mitchdenny commented on Directory.Packages.props:
```diff
@@ -120,6 +120,7 @@
     <PackageVersion Include="StackExchange.Redis" Version="2.7.33" />
     <PackageVersion Include="System.IO.Hashing" Version="8.0.0" />
     <PackageVersion Include="Yarp.ReverseProxy" Version="2.1.0" />
+    <PackageVersion Include="Elasticsearch.Net" Version="7.17.5" />
```
Deprecated package?

![image](https://github.com/dotnet/aspire/assets/513398/e3a4d615-be6f-4765-95ed-3da9a04f2b7c)

Alirexaa replied:
I have decided to add a client component with the `Elastic.Clients.Elasticsearch` package soon.
I used this deprecated package because `Elastic.Clients.Elasticsearch` was not available in the mirror package source until now.

mitchdenny commented on playground/Elasticseach/Elasticseach.AppHost/Properties/launchSettings.json:
```diff
@@ -0,0 +1,29 @@
+        "DOTNET_RESOURCE_SERVICE_ENDPOINT_URL": "https://localhost:22194"
+      }
+    },
+    "http": {
```
Add another launch profile in called `generate-manifest` (take a look at the other playground projects). Be sure to run generation on the manifest and add that file to the PR.

mitchdenny commented on src/Aspire.Hosting.Elasticsearch/ElasticsearchBuilderExtensions.cs:
```diff
@@ -0,0 +1,74 @@
+    /// <param name="name">The name of the resource. This name will be used as the connection string name when referenced in a dependency.</param>
+    /// <param name="port">The host port to bind the underlying container to.</param>
+    /// <param name="password">The parameter used to provide the superuser password for the elasticsearch. If <see langword="null"/> a random password will be generated.</param>
+    /// <returns></returns>
```
Could you append at least one `<example>` after the `<returns>`.

mitchdenny commented on src/Aspire.Hosting.Elasticsearch/ElasticsearchBuilderExtensions.cs:
```diff
@@ -0,0 +1,74 @@
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the volume. Defaults to an auto-generated name based on the application and resource names.</param>
+    /// <param name="isReadOnly">A flag that indicates if this is a read-only volume.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
```
Could you append at least one `<example>` after the `<returns>`.

mitchdenny commented on src/Aspire.Hosting.Elasticsearch/ElasticsearchBuilderExtensions.cs:
```diff
@@ -0,0 +1,74 @@
+    /// <param name="source">The source directory on the host to mount into the container.</param>
+    /// <param name="isReadOnly">A flag that indicates if this is a read-only mount.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<ElasticsearchResource> WithDataBindMount(this IResourceBuilder<ElasticsearchResource> builder, string source, bool isReadOnly = false)
```
Could you append at least one `<example>` after the `<returns>`.

mitchdenny committed:
Update ElasticsearchBuilderExtensions.cs

mitchdenny commented on tests/Aspire.Hosting.Tests/Elasticsearch/AddElasticsearchTests.cs:
```diff
@@ -0,0 +1,238 @@
+              "env": {
+                "discovery.type": "single-node",
+                "xpack.security.enabled": "true",
+                "ELASTIC_PASSWORD": "{elasticsearch-password.value}"
```
What is the deal with the username in Elasticsearch? We provide it on the connection string but it doesn't seem to be passed into the container for configuration ... is it ignored?

Alirexaa replied:
changing Elasticsearch's built-in username (like `elastic`) is impossible, but clients like `Elastic.Clients.Elasticsearch` have basic authentication using a username and password (besides using credentials in the URL).
For example :

`var settings = new ElasticsearchClientSettings(new Uri("http://localhost:9200")).Authentication(new BasicAuthentication("<USERNAME>", "<PASSWORD>"));`

so, I put the username parameter in the resource. we could remove this property from the resource and mention in the document that the username is `elastic`.

Any suggestion?

@mitchdenny , @davidfowl

mitchdenny replied:
If the username is always hardcoded (truly) to elastic then we can make it constant. No need to generate a parameter for something that could only be the one value (and people could type in the wrong thing leading to issues).

mitchdenny commented on tests/Aspire.Hosting.Tests/Elasticsearch/AddElasticsearchTests.cs:
```diff
@@ -0,0 +1,238 @@
+    {
+        using var appBuilder = TestDistributedApplicationBuilder.Create();
+
+        var kafka = appBuilder.AddElasticsearch("elasticsearch")
```
copy pasta

mitchdenny commented on tests/Aspire.Hosting.Tests/Elasticsearch/AddElasticsearchTests.cs:
```diff
@@ -0,0 +1,238 @@
+    {
+        using var appBuilder = TestDistributedApplicationBuilder.Create();
+
+        var kafka = appBuilder.AddElasticsearch("elasticsearch");
```
copy pasta

mitchdenny commented:
@Alirexaa could you please create a GitHub issue over on https://github.com/dotnet/docs-aspire with a article request and link it back to this PR. We want to make sure that we are bringing up the docs along with the feature changes like this one to help make your contribution more discoverable.

mitchdenny commented:
So I think on the `Aspire.Hosting.Elasticsearch` side things are looking pretty good (open question about username?). Just need to make sure that the XML doc comments are all there.

The component side needs more work (as @davidfowl alluded to). We will probably want an `Aspire.Elasticsearch...blah` package for the non deprecated package. It looks like it supports OTEL so it will be good to verify that all plumbs up nicely to the dashboard.

Overall, it is looking pretty good.

Alirexaa committed:
fix typo,variable name,add example docs,add aspire-manifest.json

eerhardt commented on playground/Elasticsearch/Elasticsearch.AppHost/Program.cs:
```diff
@@ -0,0 +1,21 @@
+
+var builder = DistributedApplication.CreateBuilder(args);
+
+var elasticsearch = builder.AddElasticsearch("elasticsearch")
```
`Elasticsearch` hurts my eyes, but I guess that's what the product is stylized as.

Alirexaa replied:
if you mean `Elasticsearch ` and `ElasticSearch`, yes `Elasticsearch` is the correct one but it's a little annoying.

eerhardt commented on playground/Elasticsearch/Elasticsearch.ApiService/Program.cs:
```diff
@@ -0,0 +1,39 @@
+    return response.Body;
+});
+
+app.MapPost("/", async (ElasticLowLevelClient elasticClient) =>
```
It's easier if these are all `Get`s, so you can use them without a dedicated HTTP endpoint tool. See 

https://github.com/dotnet/aspire/blob/87570cc97aa38e24f4eb4e1eb85b6926fd470025/playground/Qdrant/Qdrant.ApiService/Program.cs#L19

For how this is done elsewhere.

eerhardt commented on src/Aspire.Hosting.Elasticsearch/ElasticsearchContainerImageTags.cs:
```diff
@@ -0,0 +1,12 @@
+internal static class ElasticsearchContainerImageTags
+{
+    public const string Registry = "docker.io";
+    public const string Image = "elasticsearch";
```
```suggestion
    public const string Image = "library/elasticsearch";
```

We are explicit about image names.

eerhardt commented on src/Aspire.Hosting.Elasticsearch/ElasticsearchContainerImageTags.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public const string Registry = "docker.io";
+    public const string Image = "elasticsearch";
+    public const string Tag = "8.8.0";
```
1. This version is a year old. There are newer ones.
2. We try to use `major.minor` versions if possible, so we automatically get patch updates without needing to change our code. Does Elasticsearch support these kinds of tags, so this doesn't get out of date so quickly?

mitchdenny replied:
Seems so, I couldn't find any `major.minor` only tags. We'll just have to live with this. In this case though we'll want `8.8.2`

eerhardt commented on src/Aspire.Hosting.Elasticsearch/ElasticsearchResource.cs:
```diff
@@ -0,0 +1,60 @@
+    //All client libraries will use this port to talk to Elasticsearch
+    internal const string PrimaryEndpointName = "http";
+
+    //this endpoit is a custom binary protocol used for communications between nodes in a cluster.
```
```suggestion
    //this endpoint is a custom binary protocol used for communications between nodes in a cluster.
```

eerhardt commented on src/Aspire.Hosting.Elasticsearch/ElasticsearchResource.cs:
```diff
@@ -0,0 +1,60 @@
+    /// </summary>
+    public EndpointReference PrimaryEndpoint => _primaryEndpoint ??= new(this, PrimaryEndpointName);
+
+    /// <summary>
+    /// Gets the internal endpoint for the Elasticsearch.This endpoit used for communications between nodes in a cluster
+    /// </summary>
+    public EndpointReference InternalEndpoint => _internalEndpoint ??= new(this, InternalEndpointName);
```
Do we need this endpoint? What is the user scenario for it to be exposed in Aspire?

Alirexaa replied:
I don't have use case in my mind.
I did this as same as kafka hosting componet. But still i think we should expose internal port in publish mode.

Alirexaa committed:
Update src/Aspire.Hosting.Elasticsearch/ElasticsearchContainerImageTags.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

Alirexaa committed:
Update src/Aspire.Hosting.Elasticsearch/ElasticsearchResource.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

Alirexaa committed:
Change Elasticsearch username to constant

Alirexaa committed:
Update Elasticsearch playground

Alirexaa committed:
Update Elasticseach playground aspire-manifest

Alirexaa commented:
@mitchdenny 

please run the package import `AspNetCore.HealthChecks.Elasticsearch`

mitchdenny commented:
Will do as soon as I'm back at my desk

Alirexaa committed:
Merge branch 'dotnet:main' into elasticsearch-hosing

Alirexaa committed:
Add Aspire.Elastic.Clints.Elasticsearch component

Alirexaa commented:
After [Merge branch 'dotnet:main' into elasticsearch-hosing](https://github.com/dotnet/aspire/pull/4430/commits/cb16e86a0bf920cda289c8d68dd6ea22a0a49212)
I got this error, what is wrong?

```
Severity	Code	Description	Project	File	Line	Suppression State	Details
Error (active)	NU1101	Unable to find package Aspire.Elastic.Clients.Elasticsearch. No packages exist with this id in source(s): dotnet-eng, dotnet-public, dotnet8, dotnet9-transport. PackageSourceMapping is enabled, the following source(s) were not considered: C:\Program Files\dotnet\library-packs, dotnet-libraries.	Aspire.Elastic.Clients.Elasticsearch	D:\Workspace\aspire\src\Components\Aspire.Elastic.Clients.Elasticsearch\Aspire.Elastic.Clients.Elasticsearch.csproj	1
```
@mitchdenny @davidfowl

Alirexaa committed:
add README.md

Alirexaa committed:
fix readme

Alirexaa committed:
using Aspire.Elastic.Clients.Elasticsearch in playground

Alirexaa committed:
fix build error , add health check

Alirexaa committed:
Merge branch 'main' into elasticsearch-hosing

mitchdenny commented:
/azp run

mitchdenny commented:
> @mitchdenny
> 
> please run the package import `AspNetCore.HealthChecks.Elasticsearch`

imported

Alirexaa committed:
using PackageReference `AspNetCore.HealthChecks.Elasticsearch` with `PrivateAssets=All` in `Aspire.Elastic.Clients.Elasticsearch`

Alirexaa committed:
Merge branch 'elasticsearch-hosing' of https://github.com/Alirexaa/aspire into elasticsearch-hosing

Alirexaa committed:
add tests project

Alirexaa committed:
add more tests

Alirexaa committed:
change ManifestGenerationTests

Alirexaa committed:
add tracing

Alirexaa commented:
@mitchdenny
please run the package import `Testcontainers.Elasticsearch`

Alirexaa commented:
I have a plan to add Kibana.

I have two choices:
1- Add `WithKibana` in `ElasticsearchBuilderExtensions`
```c#

 public static IResourceBuilder<ElasticsearchResource> WithKibana(this IResourceBuilder<ElasticsearchResource> builder, Action<IResourceBuilder<KibanaResource>>? configureContainer = null, string? containerName = null)
``` 
2- `AddKibana` as separate resource hosting like `Aspire.Hosting.Kibana`

@mitchdenny

Alirexaa commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/Aspire.Elastic.Clients.Elasticsearch.csproj:
```diff
@@ -0,0 +1,29 @@
+    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" />
+    <PackageReference Include="OpenTelemetry.Extensions.Hosting" />
+
+    <!--AspNetCore.HealthChecks.Elasticsearch 8.0.1 using deprecated Nest package internaly
+    using so set 'PrivateAssets' to 'All' to prevent using ElasticClient in Nest
+    -->
+    <PackageReference Include="AspNetCore.HealthChecks.Elasticsearch" PrivateAssets="All" />
```
I will write the health check locally

Alirexaa replied:
https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/issues/2236

Alirexaa commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/AspireElasticClientsElasticsearchExtensions.cs:
```diff
@@ -0,0 +1,154 @@
+        {
+            var healthCheckName = serviceKey is null ? "Elastic.Clients.Elasticsearch" : $"Elastic.Clients.Elasticsearch_{connectionName}";
+
+            //todo: how to add health check for cloud ?
```
Now we want write Health Check locally, it's seems we can write health check for elastic cloud. Right?

Alirexaa replied:
https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/issues/2242

Alirexaa commented on src/Aspire.Hosting.Elasticsearch/ElasticsearchContainerImageTags.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public const string Registry = "docker.io";
+    public const string Image = "library/elasticsearch";
+    public const string Tag = "8.14.0";
```
https://github.com/elastic/elasticsearch/issues/110014

Alirexaa replied:
@eerhardt, please take a look https://github.com/elastic/elasticsearch/issues/110014#issuecomment-2205387000

Alirexaa commented:
> I have a plan to add Kibana.
> 
> I have two choices: 1- Add `WithKibana` in `ElasticsearchBuilderExtensions`
> 
> ```cs
>  public static IResourceBuilder<ElasticsearchResource> WithKibana(this IResourceBuilder<ElasticsearchResource> builder, Action<IResourceBuilder<KibanaResource>>? configureContainer = null, string? containerName = null)
> ```
> 
> 2- `AddKibana` as separate resource hosting like `Aspire.Hosting.Kibana`
> 
> @mitchdenny, which one is the better choice?

ping @mitchdenny, @davidfowl

mitchdenny commented:
Hrm. Interesting question. I think if Kibana is likely to be another resource you want to manage independently when it is deployed then it should be a separate resource.

Things like WithPgadmin and WithMongoExpress are more about inner loop.

Alirexaa committed:
fix HealthCheck

Alirexaa committed:
Merge branch 'main' into elasticsearch-hosing

mitchdenny commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/Guard.cs:
```diff
@@ -0,0 +1,34 @@
+    //
+    //   paramName:
+    //     The name of the parameter with which argument corresponds.
+    public static T ThrowIfNull<T>([NotNull] T? argument, bool throwOnEmptyString = false, [CallerArgumentExpression("argument")] string? paramName = null) where T : class
```
This seems unncessary.

mitchdenny commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/ElasticsearchOptions.cs:
```diff
@@ -0,0 +1,116 @@
+
+    public ElasticsearchOptions UseBasicAuthentication(string name, string password)
+    {
+        UserName = Guard.ThrowIfNull(name);
```
UserName = name ?? ArgumentNullException.ThrowIfNull(name);

mitchdenny commented:
@Alirexaa by the time you read this `Testcontainers.Elasticsearch` should be imported.

Alirexaa committed:
remove guard

Alirexaa committed:
Merge branch 'elasticsearch-hosing' of https://github.com/Alirexaa/aspire into elasticsearch-hosing

Alirexaa committed:
xml docs

Alirexaa committed:
fix xml docs

Alirexaa committed:
fix build

eerhardt commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/Aspire.Elastic.Clients.Elasticsearch.csproj:
```diff
@@ -0,0 +1,29 @@
+    <TargetFramework>$(NetCurrent)</TargetFramework>
+    <IsPackable>true</IsPackable>
+    <PackageTags>$(ComponentCommonPackageTags) Elasticsearch</PackageTags>
+    <Description>A Elasticsearch client that integrates with Aspire, including health checks, logging, and telemetry.</Description>
```
Is there an icon/logo we can use for these 2 new packages?

eerhardt replied:
Also, can you update the following files:

* https://github.com/dotnet/aspire/blob/main/src/Components/Aspire_Components_Progress.md
* https://github.com/dotnet/aspire/blob/main/src/Components/Telemetry.md

eerhardt commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/Aspire.Elastic.Clients.Elasticsearch.csproj:
```diff
@@ -0,0 +1,29 @@
+    <IsPackable>true</IsPackable>
+    <PackageTags>$(ComponentCommonPackageTags) Elasticsearch</PackageTags>
+    <Description>A Elasticsearch client that integrates with Aspire, including health checks, logging, and telemetry.</Description>
+    <NoWarn>$(NoWarn)</NoWarn>
```
```suggestion
```

This is unnecessary.

eerhardt commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/AspireElasticClientsElasticsearchExtensions.cs:
```diff
@@ -0,0 +1,153 @@
+    /// <exception cref="InvalidOperationException">Thrown when mandatory <see cref="ElasticClientsElasticsearchSettings.ConnectionString"/> is not provided.</exception>
+    /// <exception cref="InvalidOperationException">Thrown when mandatory <see cref="ElasticClientsElasticsearchSettings.Cloud"/> is not provided and <see cref="ElasticClientsElasticsearchSettings.UseCloud" /> is true .</exception>
+    public static void AddKeyedElasticClientsElasticsearch(
+      this IHostApplicationBuilder builder,
+      string name,
+      Action<ElasticClientsElasticsearchSettings>? configureSettings = null,
+      Action<ElasticsearchClientSettings>? configureClientSettings = null)
```
(nit)
```suggestion
        this IHostApplicationBuilder builder,
        string name,
        Action<ElasticClientsElasticsearchSettings>? configureSettings = null,
        Action<ElasticsearchClientSettings>? configureClientSettings = null)
```

eerhardt commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/AspireElasticClientsElasticsearchExtensions.cs:
```diff
@@ -0,0 +1,153 @@
+
+        configureSettings?.Invoke(settings);
+
+        var optionsName = serviceKey is null ? Options.Options.DefaultName : connectionName;
```
Is this used?

Alirexaa replied:
No, it was removed.

eerhardt commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/AspireElasticClientsElasticsearchExtensions.cs:
```diff
@@ -0,0 +1,153 @@
+    /// <param name="configureClientSettings">An optional delegate that can be used for customizing ElasticsearchClientSettings.</param>
+    /// <exception cref="InvalidOperationException">Thrown when mandatory <see cref="ElasticClientsElasticsearchSettings.ConnectionString"/> is not provided.</exception>
+    /// <exception cref="InvalidOperationException">Thrown when mandatory <see cref="ElasticClientsElasticsearchSettings.Cloud"/> is not provided and <see cref="ElasticClientsElasticsearchSettings.UseCloud" /> is true .</exception>
+    public static void AddElasticClientsElasticsearch(
```
```suggestion
    public static void AddElasticsearchClient(
```

This reads better and follows our naming pattern used elsewhere.

eerhardt commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/AspireElasticClientsElasticsearchExtensions.cs:
```diff
@@ -0,0 +1,153 @@
+    /// <param name="configureClientSettings">An optional delegate that can be used for customizing ElasticsearchClientSettings.</param>
+    /// <exception cref="InvalidOperationException">Thrown when mandatory <see cref="ElasticClientsElasticsearchSettings.ConnectionString"/> is not provided.</exception>
+    /// <exception cref="InvalidOperationException">Thrown when mandatory <see cref="ElasticClientsElasticsearchSettings.Cloud"/> is not provided and <see cref="ElasticClientsElasticsearchSettings.UseCloud" /> is true .</exception>
+    public static void AddKeyedElasticClientsElasticsearch(
```
```suggestion
    public static void AddKeyedElasticsearchClient(
```

eerhardt commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/AssemblyInfo.cs:
```diff
@@ -0,0 +1,9 @@
+using Aspire;
+using Aspire.Elastic.Clients.Elasticsearch;
+
+[assembly: ConfigurationSchema("Aspire.Elastic.Clients.Elasticsearch", typeof(ElasticClientsElasticsearchSettings))]
```
```suggestion
[assembly: ConfigurationSchema("Aspire:Elastic:Clients:Elasticsearch", typeof(ElasticClientsElasticsearchSettings))]
```

eerhardt commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/AspireElasticClientsElasticsearchExtensions.cs:
```diff
@@ -0,0 +1,153 @@
+
+        configureClientSettings?.Invoke(elasticsearchClientSettings);
+
+        var elasticsearchClient = new ElasticsearchClient(elasticsearchClientSettings);
```
Do we want to delay creating this object until it is needed/resolved from DI?

eerhardt commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/ElasticsearchHealthCheckBuilderExtensions.cs:
```diff
@@ -0,0 +1,125 @@
+    /// <param name="tags">A list of tags that can be used to filter sets of health checks. Optional.</param>
+    /// <param name="timeout">An optional <see cref="TimeSpan"/> representing the timeout of the check.</param>
+    /// <returns>The specified <paramref name="builder"/>.</returns>
+    public static IHealthChecksBuilder AddElasticsearch(
```
Can we trim the unused code? Hopefully the new HealthCheck package is released soon and we can just use that.

@adamsitnik @unaizorrilla @sungam3r - any thoughts on when a new https://www.nuget.org/packages/AspNetCore.HealthChecks.Elasticsearch version can be released with https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/2244?

adamsitnik replied:
Hmm since it's a breaking change, we would need to release first `9.0-*` package (rather than just 8.0.2). To do that we should most likely update all MS and System dependencies to `9.0-`. It's not hard, but time consuming (it should take less time than last year with https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/2044 since I've refactored the yaml files).

@eerhardt how important is this to you? If it's not very important, I would prefer to wait for RC1 bits being available at nuget.org.

cc @unaizorrilla

eerhardt replied:
> To do that we should most likely update all MS and System dependencies to 9.0-.

We should have a conversation about that. Since .NET 9 isn't LTS, we don't want to force Aspire users to move to net9.0 TFM just to take advantage of new Aspire features. The current plan (from my understanding) is that Aspire 9.0 will target net8.0 TFM, so users can use either net8.0 or net9.0 in their projects. This is similar to how efcore operates - it targets the last LTS.

* https://www.nuget.org/packages/Microsoft.EntityFrameworkCore/9.0.0-preview.6.24327.4#dependencies-body-tab
* https://www.nuget.org/packages/Microsoft.EntityFrameworkCore/8.0.7#dependencies-body-tab
* https://www.nuget.org/packages/Microsoft.EntityFrameworkCore/7.0.20#dependencies-body-tab
* https://www.nuget.org/packages/Microsoft.EntityFrameworkCore/6.0.32#dependencies-body-tab

Notice that `9.0.0-preview.6.24327.4` and `8.0.7` target `net8.0` while `7.0.20` and `6.0.32` target `net6.0`.

eerhardt commented on tests/Aspire.Elastic.Clients.Elasticsearch.Tests/README.md:
```diff
@@ -0,0 +1,3 @@
+# Aspire.Elastic.Clients.Elasticsearch.Tests
```
I think this file can be deleted. It doesn't add much value.

Alirexaa committed:
Apply suggestions from code review

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

Alirexaa committed:
fix compile errors

Alirexaa committed:
delay creating Elasticsearch client

Alirexaa committed:
remove HealthCheck unused code

Alirexaa committed:
clean renaming AddElasticClientsElasticsearch -> AddElasticsearchClient

Alirexaa committed:
remove readme in test project

mitchdenny commented on NuGet.config:
```diff
@@ -17,6 +17,7 @@
     <add key="dotnet8" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json" />
     <add key="dotnet-libraries" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-libraries/nuget/v3/index.json" />
     <add key="dotnet9-transport" value="https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet9-transport/nuget/v3/index.json" />
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
```
All dependencies should be here now. Time to revert?

mitchdenny committed:
Remove isReadonly parameters.

mitchdenny commented:
@Alirexaa this is looking good. I removed the `isReadonly` parameters since when I set it to readonly Elasticsearch failed to start up properly so only imply that it does work (when it doesn't).

One final thing that I'd like to see done before I approve this is splitting the tests out of `Aspire.Hosting.Tests` into a new test assembly named `Aspire.Hosting.Elasticsearch.Tests`. We are starting the process of breaking up tests into different assemblies.

We've done this in `Aspire.Hosting.Containers.Tests` if you want an example.

eerhardt committed:
Minor PR feedback

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into elasticsearch-hosing

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into elasticsearch-hosing

eerhardt commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/ElasticClientsElasticsearchSettings.cs:
```diff
@@ -0,0 +1,52 @@
+    /// <remarks>
+    /// Set value <see langword="true"/> when you want to use elastic cloud and provide api key and elasticId in <see cref="ElasticClientsElasticsearchSettings.Cloud"/>.
+    /// </remarks>
+    public bool UseCloud { get; set; }
```
Do we need this `bool` property? Or can we just use the fact that if `ElasticClientsElasticsearchCloudSettings? Cloud` is non-null, then you want to use the elastic cloud.

eerhardt commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/AssemblyInfo.cs:
```diff
@@ -0,0 +1,7 @@
+using Aspire;
+using Aspire.Elastic.Clients.Elasticsearch;
+
+[assembly: ConfigurationSchema("Aspire:Elastic:Clients:Elasticsearch", typeof(ElasticClientsElasticsearchSettings))]
```
Note that I removed the LoggingCategory because I didn't see anything in the library that would log to the ILogger. Am I missing something? Or is there no logging support for ElasticsearchClient.

eerhardt reviewed (commented):
Thanks for the contribution @Alirexaa. I think this PR is getting close to being able to approval.

I merged this PR with main and made a few minor tweaks to the code.

radical commented:
I have mirrored the `Testcontainers.Elasticsearch` package, and re-run the build.

radical commented on tests/Aspire.Elastic.Clients.Elasticsearch.Tests/AspireElasticClientExtensionsTest.cs:
```diff
@@ -0,0 +1,146 @@
+        builder.AddKeyedElasticsearchClient(key, settings =>
+        {
+            settings.DisableHealthChecks = false;
+            settings.HealthCheckTimeout = 1;
```
Just curious - why is this required?

radical commented on tests/Aspire.Elastic.Clients.Elasticsearch.Tests/AspireElasticClientExtensionsTest.cs:
```diff
@@ -0,0 +1,146 @@
+
+    [Fact]
+    [RequiresDocker]
+    public async Task AddKeyedElasticsearchClient_HealthCheckShouldBeRegisteredWhenEnabled()
```
The keyed, and non-keyed versions of the tests can be combined like:
https://github.com/dotnet/aspire/blob/3406514607bb775cb5c9693c269832ca5c5a4170/tests/Aspire.NATS.Net.Tests/AspireNatsClientExtensionsTests.cs#L121-L126

radical commented on tests/Aspire.Elastic.Clients.Elasticsearch.Tests/AspireElasticClientExtensionsTest.cs:
```diff
@@ -0,0 +1,146 @@
+
+namespace Aspire.Elastic.Clients.Elasticsearch.Tests;
+
+public class AspireElasticClientExtensionsTest : IClassFixture<ElasticsearchContainerFixture>
```
Can we also add a test similar to this? https://github.com/dotnet/aspire/blob/3406514607bb775cb5c9693c269832ca5c5a4170/tests/Aspire.NATS.Net.Tests/AspireNatsClientExtensionsTests.cs#L178-L204

@eerhardt would that make sense here?

eerhardt replied:
Yes, I think that would be a good to have this test.

radical commented on tests/Aspire.Elastic.Clients.Elasticsearch.Tests/ConformanceTests.cs:
```diff
@@ -0,0 +1,80 @@
+
+        configuration.AddInMemoryCollection(
+            [
+            new KeyValuePair<string, string?>(CreateConfigKey("Aspire:Elastic:Clients:Elasticsearch", key, "ConnectionString"),connectionString),
+            new KeyValuePair<string, string?>($"ConnectionStrings:{key}",connectionString)
```
```suggestion
                new KeyValuePair<string, string?>(CreateConfigKey("Aspire:Elastic:Clients:Elasticsearch", key, "ConnectionString"), connectionString),
                new KeyValuePair<string, string?>($"ConnectionStrings:{key}", connectionString)
```

radical commented on src/Aspire.Hosting.Elasticsearch/ElasticsearchBuilderExtensions.cs:
```diff
@@ -0,0 +1,112 @@
+        IResourceBuilder<ParameterResource>? password = null,
+        int? port = null)
+    {
+
```
nit:
```suggestion
```

radical commented on src/Aspire.Hosting.Elasticsearch/ElasticsearchResource.cs:
```diff
@@ -0,0 +1,57 @@
+        PasswordParameter = password;
+    }
+
+    private EndpointReference? _primaryEndpoint;
+    private EndpointReference? _internalEndpoint;
+
+    /// <summary>
+    /// Gets the primary endpoint for the Elasticsearch. This endpoint is used for all API calls over HTTP.
+    /// </summary>
+    public EndpointReference PrimaryEndpoint => _primaryEndpoint ??= new(this, PrimaryEndpointName);
+
+    /// <summary>
+    /// Gets the internal endpoint for the Elasticsearch. This endpoint used for communications between nodes in a cluster
+    /// </summary>
+    public EndpointReference InternalEndpoint => _internalEndpoint ??= new(this, InternalEndpointName);
```
Just curious - would it make sense to init the properties in the constructor?

radical commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/AspireElasticClientsElasticsearchExtensions.cs:
```diff
@@ -0,0 +1,149 @@
+    /// <param name="builder">The <see cref="IHostApplicationBuilder" /> to read config from and add services to.</param>
+    /// <param name="connectionName">A name used to retrieve the connection string from the ConnectionStrings configuration section.</param>
+    /// <param name="configureSettings">An optional delegate that can be used for customizing options. It's invoked after the settings are read from the configuration.</param>
+    /// <param name="configureClientSettings">An optional delegate that can be used for customizing ElasticsearchClientSettings.</param>
```
Should there be a cref for `ElasticsearchClientSettings` also?

radical commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/AspireElasticClientsElasticsearchExtensions.cs:
```diff
@@ -0,0 +1,149 @@
+        string connectionName,
+        Action<ElasticClientsElasticsearchSettings>? configureSettings = null,
+        Action<ElasticsearchClientSettings>? configureClientSettings = null
+        ) => builder.AddElasticsearchClient(DefaultConfigSectionName, configureSettings, configureClientSettings, connectionName, null);
```
nit:
```suggestion
        ) => builder.AddElasticsearchClient(DefaultConfigSectionName, configureSettings, configureClientSettings, connectionName, serviceKey: null);
```

radical commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/AspireElasticClientsElasticsearchExtensions.cs:
```diff
@@ -0,0 +1,149 @@
+    {
+        if (settings.UseCloud)
+        {
+            if (settings.Cloud is null)
+            {
+                throw new InvalidOperationException($"No Cloud specified. Ensure a valid cloudId and api key was provided in '{configurationSectionName}:Cloud' configuration key.");
+            }
+            if (settings.Cloud.CloudId is null)
+            {
+                throw new InvalidOperationException($"No CloudId specified. Ensure a valid cloudId provided in '{configurationSectionName}:Cloud:CloudId' configuration key.");
+            }
+            if (settings.Cloud.ApiKey is null)
+            {
+                throw new InvalidOperationException($"No ApiKey specified. Ensure a valid api key provided in '{configurationSectionName}:Cloud:ApiKey' configuration key.");
+            }
```
Are there tests covering these?

radical commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/AspireElasticClientsElasticsearchExtensions.cs:
```diff
@@ -0,0 +1,149 @@
+        {
+            if (settings.ConnectionString is null)
+            {
+                throw new InvalidOperationException($"No ConnectionString specified. Ensure a valid connection string was provided in 'ConnectionStrings:{connectionName}' or for the '{configurationSectionName}:ConnectionString' configuration key.");
```
.. and test for this.

radical commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/ElasticClientsElasticsearchSettings.cs:
```diff
@@ -0,0 +1,52 @@
+    public int? HealthCheckTimeout { get; set; }
+
+    /// <summary>
+    /// Gets or sets a boolean value that indicates whether the connection to elastic cloud or not.
```
```suggestion
    /// Gets or sets a boolean value that indicates whether the connection should use the elastic cloud or not.
```

radical commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/ElasticsearchHealthCheck.cs:
```diff
@@ -0,0 +1,37 @@
+using Elastic.Clients.Elasticsearch;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+
+// TODO: Use health check from AspNetCore.Diagnostics.HealthChecks once following PR released:
+// https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/2244
```
Can we open an issue to track that this needs to be done, if the nuget isn't available yet?

radical commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/ElasticsearchHealthCheck.cs:
```diff
@@ -0,0 +1,37 @@
+// TODO: Use health check from AspNetCore.Diagnostics.HealthChecks once following PR released:
+// https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/2244
+namespace Aspire.Elastic.Clients.Elasticsearch;
+internal sealed class ElasticsearchHealthCheck : IHealthCheck
```
```suggestion

internal sealed class ElasticsearchHealthCheck : IHealthCheck
```

radical commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/README.md:
```diff
@@ -0,0 +1,147 @@
+    settings.UseCloud = true;
+    settings.Cloud.CloudId = "Valid CloudId";
+    settings.Cloud.ApiKey = "Valid ApiKey";
+
```
```suggestion
```

radical commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/README.md:
```diff
@@ -0,0 +1,147 @@
+when calling `builder.AddElasticsearchClient()`.
+Example appsettings.json that configures the options:
+
+
```
```suggestion
```

mitchdenny commented:
Importing updated package for `Testcontainers.Elasticsearch`.

mitchdenny reviewed (approved):
Looking good from my perspective. Would like to see the nits that others have addressed but overall I think this is a good to go in for 8.1

radical commented:
> Importing updated package for `Testcontainers.Elasticsearch`.

I imported that, but forgot to trigger a re-run here.

stevejgordon reviewed (commented):
@Alirexaa This is amazing, thank you. It's exciting to see ES support coming to Aspire. It looks and works great! My only suggestion for how it could be enhanced (in the future) is to consider supporting running Kibana as well. That might be more involved with the container communication. It's common for developers to use Kibana to visually verify the data. 

Thanks again for contributing this amazing PR.

Mpdreamz commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/AspireElasticClientsElasticsearchExtensions.cs:
```diff
@@ -0,0 +1,149 @@
+        string connectionName,
+        string configurationSectionName)
+    {
+        if (settings.UseCloud)
```
This `UseCloud` configuration feels superfluous perhaps? Setting `settings.Cloud` to anything but `null` signals the same intent.

Alirexaa commented:
@mitchdenny, before merging PR, let me make some suggested changes.

Mpdreamz commented on src/Components/Aspire.Elastic.Clients.Elasticsearch/ConfigurationSchema.json:
```diff
@@ -0,0 +1,63 @@
+                    "Cloud": {
+                      "type": "object",
+                      "properties": {
+                        "ApiKey": {
```
ApiKey authentication is not tied to Elastic Cloud, you can use API keys on self hosted Elasticsearch too:

https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-create-api-key.html

Maybe create a separate `Authentication` configuration object that supports ApiKey and basic auth?

Alirexaa replied:
In my opinion `Aspire.Elastic.Clients.Elasticsearch` is built for use in Aspire apps and `ElasticsearchResource` that is self-hosted uses password authentication right now, so if we want to support ApiKey in the client component, we should support ApiKey in hosting component too, right? 

I am not sure but I think if we want to support ApiKey in `ElasticsearchResource` we may face a problem in 
creating ConnectionString.

eerhardt replied:
The pattern we've been following in Aspire for connection information is to encode all the information needed to establish the connection into the `ConnectionString:<name>` configuration value, and then parsing the connection information from that. For example see:

https://github.com/dotnet/aspire/blob/6a4610c1d0d8213ecb3e5d0465256434dcb60852/src/Components/Aspire.Qdrant.Client/QdrantClientSettings.cs#L26-L47

https://github.com/dotnet/aspire/blob/6a4610c1d0d8213ecb3e5d0465256434dcb60852/src/Components/Aspire.Azure.AI.OpenAI/AzureOpenAISettings.cs#L52-L72

Is this a pattern we can follow here? So these "Cloud" connection properties can be parsed from that string? Then in order to switch, you just need to update that string.

Alirexaa replied:
> Is this a pattern we can follow here? So these "Cloud" connection properties can be parsed from that string? Then in order to switch, you just need to update that string.

Do you mean remove the `Cloud` section from `ElasticClientsElasticsearchSettings` or keep `Cloud` and retrieve ApiKey and CloudId from ConnectionString when the `Cloud` section is null?

by encoding information in ConnectionString we could support ApiKey in `ElasticsearchResource` too, it would be something like `Endpoint=Host:Port,ApiKey=someApiKey`, but do we need to support it in this PR?

eerhardt replied:
I think we could rework the Settings class a bit.

```csharp
public sealed class ElasticClientsElasticsearchSettings
{
    public Uri? Endpoint { get; set; }
    public string? Key { get; set; } // or ApiKey
    public string? CloudId { get; set; }
...
}
```

And then following the same pattern as shown in the above 2 existing libraries for breaking apart the `ConnectionStrings:<name>` string into these properties.

Then the question becomes what happens when both `Endpoint` and `CloudId` are specified? We can either throw an exception, or say one of them wins over the other.

> but do we need to support it in this PR?

I think we should get the initial design right, so we don't have to make potentially breaking changes in the future.

eerhardt replied:
> but do we need to support it in this PR?

To be clear - I don't think we need to support more scenarios than what is currently supported in this PR. I just want to make sure the public API is set up for future enhancements.

eerhardt replied:
> ApiKey authentication is not tied to Elastic Cloud, you can use API keys on self hosted Elasticsearch too

@Mpdreamz @stevejgordon - how is this possible with the .NET client? I only see these constructors:

* public ElasticsearchClientSettings(Uri? uri = null)
* public ElasticsearchClientSettings(string cloudId, AuthorizationHeader credentials)

How do you create an `ElasticsearchClient` using an API key on self hosted Elasticsearch?

Mpdreamz replied:
Since cloud requires it its part of the constructor but otherwise you can call `. Authentication()` on the settings itself for other scenarios.

Mpdreamz replied:
I Like the simplification of the settings so it projects 1 to 1 to a connection string.

I do think a single endpoint might not be sufficiently future proof for production deployments:

Scenario's

- Elastic CloudId. 
- Endpoints: []
- UseSniffing. 

If `Endpoint` is a SAAS offering a single one will be sufficient. 
However if it points to a self hosted Elasticsearch deployment we might want to support multiple endpoints.

Obviously in dev you would typically only launch a single Elasticsearch instance but in production you often have `>3` nodes to ensure availability. 

Elastic's transport layer can be seeded with multiple endpoints (e.g the master eligible nodes). Using sniffing the transport [can discover all the endpoints](https://github.com/elastic/elastic-transport-net/blob/eb5416e8df75e4c50215bf5258e13189afd9017d/src/Elastic.Transport/Products/Elasticsearch/ElasticsearchProductRegistration.cs#L103) that have data to be queried. This ensures the transport does not create hotspots and that it leaves nodes alone that should not be queried directly. See https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-node.html for more about node roles. This way we play nicely with medium to large size deployments as well.

We could simplify this by always enabling sniffing OOTB if multiple endoints are specified.

Alirexaa committed:
Apply suggestions from code review

Co-authored-by: Ankit Jain <radical@gmail.com>

Alirexaa committed:
seperate Elaticsearch tests

Alirexaa committed:
fix build error

radical commented:
Merge `main` to pick up fixes for `Aspire.Hosting.Containers.Tests` failing here.

Alirexaa committed:
Merge branch 'main' into elasticsearch-hosing

Alirexaa committed:
add suggested changes

Alirexaa committed:
Merge branch 'elasticsearch-hosing' of https://github.com/Alirexaa/aspire into elasticsearch-hosing

radical commented:
Merge `main` again to pick up fixes for `Aspire.Hosting.Redis.Tests`.

Alirexaa committed:
Merge branch 'dotnet:main' into elasticsearch-hosing

radical commented:
There are a few relevant test failures [here](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-aspire-refs-pull-4430-merge-4d077d62154540d9a8/Aspire.Elastic.Clients.Elasticsearch.Tests/1/console.1395d841.log?helixlogtype=result).

```
  Failed Aspire.Elastic.Clients.Elasticsearch.Tests.AspireElasticClientExtensionsTest.ThrowInvalidOperationExceptionWhenConfigurationNotProvided(useKeyed: True) [79 ms]
  Error Message:
   Assert.Throws() Failure
Expected: typeof(System.InvalidOperationException)
Actual:   (No exception was thrown)
  Stack Trace:
     at Aspire.Elastic.Clients.Elasticsearch.Tests.AspireElasticClientExtensionsTest.ThrowInvalidOperationExceptionWhenConfigurationNotProvided(Boolean useKeyed) in /_/tests/Aspire.Elastic.Clients.Elasticsearch.Tests/AspireElasticClientExtensionsTest.cs:line 149
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
  Failed Aspire.Elastic.Clients.Elasticsearch.Tests.ConformanceTests.HealthCheckReportsExpectedStatus(key: "key") [4 s]
  Error Message:
   Assert.Equal() Failure
Expected: Healthy
Actual:   Unhealthy
  Stack Trace:
     at Aspire.Components.ConformanceTests.ConformanceTests`2.HealthCheckReportsExpectedStatus(String key) in /_/tests/Aspire.Components.Common.Tests/ConformanceTests.cs:line 314
--- End of stack trace from previous location ---
  Failed Aspire.Elastic.Clients.Elasticsearch.Tests.AspireElasticClientExtensionsTest.ThrowInvalidOperationExceptionWhenConfigurationNotProvided(useKeyed: False) [< 1 ms]
  Error Message:
   Assert.Throws() Failure
Expected: typeof(System.InvalidOperationException)
Actual:   (No exception was thrown)
  Stack Trace:
     at Aspire.Elastic.Clients.Elasticsearch.Tests.AspireElasticClientExtensionsTest.ThrowInvalidOperationExceptionWhenConfigurationNotProvided(Boolean useKeyed) in /_/tests/Aspire.Elastic.Clients.Elasticsearch.Tests/AspireElasticClientExtensionsTest.cs:line 153
   at InvokeStub_AspireElasticClientExtensionsTest.ThrowInvalidOperationExceptionWhenConfigurationNotProvided(Object, Span`1)
   at System.Reflection.MethodBaseInvoker.InvokeWithOneArg(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
  Failed Aspire.Elastic.Clients.Elasticsearch.Tests.ConformanceTests.HealthCheckReportsExpectedStatus(key: null) [143 ms]
  Error Message:
   Assert.Equal() Failure
Expected: Healthy
Actual:   Unhealthy
  Stack Trace:
     at Aspire.Components.ConformanceTests.ConformanceTests`2.HealthCheckReportsExpectedStatus(String key) in /_/tests/Aspire.Components.Common.Tests/ConformanceTests.cs:line 314
--- End of stack trace from previous location ---
```

radical commented:
The EndToEnd tests are failing because we need to add an entry for the new package at https://github.com/eerhardt/aspire/blob/08c3553a4c5a1f2c45ce94957a359b5a8bffd3af/tests/testproject/Directory.Packages.Helix.props#L51-L52

sebastienros committed:
Add Aspire.Hosting.Elasticsearch to Directory.Packages.Helix.props

sebastienros committed:
Merge branch 'main' into elasticsearch-hosing

Alirexaa committed:
remove invalid test

Alirexaa committed:
fix type name

Alirexaa committed:
fix failing test

radical commented:
[This](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-aspire-refs-pull-4430-merge-a573f662f17848fe86/Aspire.EndToEnd.Tests-basicservices/1/console.8c0bef30.log?helixlogtype=result) `Aspire.EndToEnd.Tests.IntegrationServicesTests.VerifyComponentWorks(resourceName: elasticsearch) [FAIL]` is getting timed out waiting for a response.

The httpclient retry log messages show:
```
{
   "detail" : "Unsuccessful () low level call on GET: /
	 Exception: The request was canceled due to the configured HttpClient.Timeout of 60 seconds elapsing.

	# Audit trail of this API call:
	 - [1] BadRequest: Node: http://elastic:redacted@localhost:44039/ Took: 00:01:00.0047488
	 - [2] MaxTimeoutReached:
	# OriginalException: System.Threading.Tasks.TaskCanceledException: The request was canceled due to the configured HttpClient.Timeout of 60 seconds elapsing.
	 ---> System.TimeoutException: The operation was canceled.
	 ---> System.Threading.Tasks.TaskCanceledException: The operation was canceled.
	 ---> System.Net.Http.HttpIOException: The response ended prematurely. (ResponseEnded)
	   at System.Net.Http.HttpConnection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
	   --- End of inner exception stack trace ---
	   at System.Net.Http.HttpConnection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
	   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
	   at System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
	   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
	   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
	   --- End of inner exception stack trace ---
	   --- End of inner exception stack trace ---
	   at System.Net.Http.HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts)
	   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
	   at Elastic.Transport.HttpRequestInvoker.RequestCoreAsync[TResponse](Boolean isAsync, RequestData requestData, CancellationToken cancellationToken)
	# Request:
	<Request stream not captured or already read to completion by serializer. Set DisableDirectStreaming() on TransportConfiguration to force it to be set on the response.>
	# Response:
	<Response stream not captured or already read to completion by serializer. Set DisableDirectStreaming() on TransportConfiguration to force it to be set on the response.>
	",
   "status" : 500,
   "title" : "An error occurred while processing your request.",
   "type" : "https://tools.ietf.org/html/rfc9110#section-15.6.1"
}
```

The container log dumped at the end shows:
```
 [container-list] Executing - docker container list --all --filter name=elasticsearch --format {{.Names}} 
 [container-list] elasticsearch-bcpfvshn-3ca60341
 [elasticsearch] Executing - docker container logs elasticsearch-bcpfvshn-3ca60341 -n 50 
 [elasticsearch] Created elasticsearch keystore in /usr/share/elasticsearch/config/elasticsearch.keystore
 [elasticsearch] 
 [elasticsearch] ERROR: Elasticsearch died while starting up, with exit code 137
```

Another thing to note is that this test ran to the timeout for all of the tests - 15mins. So maybe we need to bump that. But that is if the container failure is irrelevant.

Alirexaa commented:
> [This](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-aspire-refs-pull-4430-merge-a573f662f17848fe86/Aspire.EndToEnd.Tests-basicservices/1/console.8c0bef30.log?helixlogtype=result) `Aspire.EndToEnd.Tests.IntegrationServicesTests.VerifyComponentWorks(resourceName: elasticsearch) [FAIL]` is getting timed out waiting for a response.
> 
> The httpclient retry log messages show:
> 
> ```
> {
>    "detail" : "Unsuccessful () low level call on GET: /
> 	 Exception: The request was canceled due to the configured HttpClient.Timeout of 60 seconds elapsing.
> 
> 	# Audit trail of this API call:
> 	 - [1] BadRequest: Node: http://elastic:redacted@localhost:44039/ Took: 00:01:00.0047488
> 	 - [2] MaxTimeoutReached:
> 	# OriginalException: System.Threading.Tasks.TaskCanceledException: The request was canceled due to the configured HttpClient.Timeout of 60 seconds elapsing.
> 	 ---> System.TimeoutException: The operation was canceled.
> 	 ---> System.Threading.Tasks.TaskCanceledException: The operation was canceled.
> 	 ---> System.Net.Http.HttpIOException: The response ended prematurely. (ResponseEnded)
> 	   at System.Net.Http.HttpConnection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
> 	   --- End of inner exception stack trace ---
> 	   at System.Net.Http.HttpConnection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
> 	   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
> 	   at System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
> 	   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
> 	   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
> 	   --- End of inner exception stack trace ---
> 	   --- End of inner exception stack trace ---
> 	   at System.Net.Http.HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts)
> 	   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
> 	   at Elastic.Transport.HttpRequestInvoker.RequestCoreAsync[TResponse](Boolean isAsync, RequestData requestData, CancellationToken cancellationToken)
> 	# Request:
> 	<Request stream not captured or already read to completion by serializer. Set DisableDirectStreaming() on TransportConfiguration to force it to be set on the response.>
> 	# Response:
> 	<Response stream not captured or already read to completion by serializer. Set DisableDirectStreaming() on TransportConfiguration to force it to be set on the response.>
> 	",
>    "status" : 500,
>    "title" : "An error occurred while processing your request.",
>    "type" : "https://tools.ietf.org/html/rfc9110#section-15.6.1"
> }
> ```
> 
> The container log dumped at the end shows:
> 
> ```
>  [container-list] Executing - docker container list --all --filter name=elasticsearch --format {{.Names}} 
>  [container-list] elasticsearch-bcpfvshn-3ca60341
>  [elasticsearch] Executing - docker container logs elasticsearch-bcpfvshn-3ca60341 -n 50 
>  [elasticsearch] Created elasticsearch keystore in /usr/share/elasticsearch/config/elasticsearch.keystore
>  [elasticsearch] 
>  [elasticsearch] ERROR: Elasticsearch died while starting up, with exit code 137
> ```
> 
> Another thing to note is that this test ran to the timeout for all of the tests - 15mins. So maybe we need to bump that. But that is if the container failure is irrelevant.

It seems container stopped.
I'm not sure but exit code 137 mean a process is terminated externally because of its memory consumption.
Do we have limit for this?
@radical

Alirexaa committed:
add functional tests

eerhardt commented on tests/Aspire.EndToEnd.Tests/IntegrationServicesTests.cs:
```diff
@@ -35,6 +35,7 @@ public IntegrationServicesTests(ITestOutputHelper testOutput, IntegrationService
     [InlineData(TestResourceNames.sqlserver)]
     [InlineData(TestResourceNames.efsqlserver)]
     [InlineData(TestResourceNames.milvus)]
+    [InlineData(TestResourceNames.elasticsearch)]
```
Now that we have "FunctionalTests" in ElasticsearchFunctionalTests, I don't think we need to add this to the EndToEnd tests.

Using the new separate Hosting tests pattern, along with functional tests there, means we don't need to have every resource/component pair in the EndToEnd tests. Instead we can just use some "core" resources and components in the EndToEnd, and the other resources can be tested in their own assembly.

cc @radical @mitchdenny @sebastienros

Alirexaa committed:
add elastic logo

Alirexaa committed:
Update Telemtry And Component Progress

eerhardt committed:
Add SuppressFinalPackageVersion for the initial version of the Elasticsearch libraries

This allows us to ship these libraries as preview even when the rest of the Aspire libraries ship stable. Will remove and ship stable once the libraries have had time to get feedback.

radical committed:
Merge branch 'main' into elasticsearch-hosing

radical commented:
Fixed the merge conflicts.

radical committed:
fix merge

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into elasticsearch-hosing

radical commented:
@eerhardt Tests failing with: `could not create the container	{"Container": {"name":"elasticsearch-ejmsbhdf-a3b368e"}, "Reconciliation": 3, "error": "docker command 'CreateContainer' returned with non-zero exit code 1: command output: Stdout: '' Stderr: 'Unable to find image 'netaspireci.azurecr.io/library/elasticsearch:8.14.0' locally\nError response from daemon: manifest for netaspireci.azurecr.io/library/elasticsearch:8.14.0 not found: manifest unknown: manifest tagged by \"8.14.0\" is not found\n'"}`

radical committed:
Merge remote-tracking branch 'origin/main' into elasticsearch-hosing

# Conflicts:
#	tests/Shared/RepoTesting/Directory.Packages.Helix.props
#	tests/testproject/TestProject.AppHost/TestProject.AppHost.csproj
#	tests/testproject/TestProject.IntegrationServiceA/TestProject.IntegrationServiceA.csproj

eerhardt committed:
Remove elasticsearch from EndToEnd tests now that we have functional tests that cover the same functionality.

eerhardt committed:
Merge remote-tracking branch 'Alirexaa/elasticsearch-hosing' into elasticsearch-hosing

eerhardt commented:
> @eerhardt Tests failing with: `could not create the container {"Container": {"name":"elasticsearch-ejmsbhdf-a3b368e"}, "Reconciliation": 3, "error": "docker command 'CreateContainer' returned with non-zero exit code 1: command output: Stdout: '' Stderr: 'Unable to find image 'netaspireci.azurecr.io/library/elasticsearch:8.14.0' locally\nError response from daemon: manifest for netaspireci.azurecr.io/library/elasticsearch:8.14.0 not found: manifest unknown: manifest tagged by \"8.14.0\" is not found\n'"}`

I've pushed the image to our mirror. This should be fixed now.

eerhardt commented:
FYI - I've removed the EndToEnd additions as they are no longer necessary now that we have functional tests for Elasticsearch.

eerhardt committed:
fix tests

mitchdenny commented:
@Alirexaa thank you once again for your contribution to .NET Aspire. This PR took a while to get merged and we appreciate you hanging in there whilst we go through multiple review iterations.

Just so you know we plan on shipping this in .NET Aspire 8.1 as a preview package to allow us to gather some broader usage feedback with a few of marking the package and API as stable in .NET Aspire 8.2 when it rolls around.

We are in the process of writing the blog post for .NET Aspire 8.2, are you happy for us to give you a shout out and link to your GitHub profile in that post?

EwoutdBoer commented:
I will be testing this (preview) package. @Alirexaa thanks for your work. I could now completely implement Aspire in my current project, while the only missing part was Elastic search.

Alirexaa commented:
> @Alirexaa thank you once again for your contribution to .NET Aspire. This PR took a while to get merged and we appreciate you hanging in there whilst we go through multiple review iterations.
> 
Glad to working here. Thanks to you and team that gave feedback to PR. 

> We are in the process of writing the blog post for .NET Aspire 8.2, are you happy for us to give you a shout out and link to your GitHub profile in that post?
Yes sure.

Alirexaa commented:
> I will be testing this (preview) package. @Alirexaa thanks for your work. I could now completely implement Aspire in my current project, while the only missing part was Elastic search.

Glad to hear. Thanks for you feedback.
